File: src/Magnifier.cpp
#include "Magnifier.h"

#include <magnification.h>
#include <windows.h>

#include <algorithm>

const float Magnifier::DEFAULT_MAG_LEVEL = 1.5;

Magnifier::Magnifier(float mag_level, int dead_zone)
    : mag_level(mag_level), dead_zone(dead_zone), magnified(false) {
    // Need some exception
    MagInitialize();

    int width = GetSystemMetrics(SM_CXSCREEN);
    int height = GetSystemMetrics(SM_CYSCREEN);

    float mul_x = width * (1 - 1 / mag_level) / (width - 2 * dead_zone);
    float mul_y = height * (1 - 1 / mag_level) / (height - 2 * dead_zone);
    float sub_x = mul_x * dead_zone;
    float sub_y = mul_y * dead_zone;
    int max_x = static_cast<int>(width * (1 - 1 / mag_level));
    int max_y = static_cast<int>(height * (1 - 1 / mag_level));

    transform_x = [mul_x, sub_x, max_x](int x) {
        return std::clamp(static_cast<int>(x * mul_x - sub_x), 0, max_x);
    };
    transform_y = [mul_y, sub_y, max_y](int y) {
        return std::clamp(static_cast<int>(y * mul_y - sub_y), 0, max_y);
    };
}

Magnifier::~Magnifier() {
    // Need some exception
    MagUninitialize();
}

void Magnifier::magnify() const {
    if (!magnified) return;

    POINT p{};
    // Need some exception
    GetCursorPos(&p);

    // Need some exception
    int x = transform_x(p.x);
    int y = transform_y(p.y);
    MagSetFullscreenTransform(mag_level, x, y);

    RECT rcDest;
    rcDest.left = 0;
    rcDest.top = 0;
    rcDest.right = GetSystemMetrics(SM_CXSCREEN);
    rcDest.bottom = GetSystemMetrics(SM_CYSCREEN);

    RECT rcSource;
    rcSource.left = x;
    rcSource.top = y;
    rcSource.right = rcSource.left + (int)(rcDest.right / mag_level);
    rcSource.bottom = rcSource.top + (int)(rcDest.bottom / mag_level);

    // Need some exception
    MagSetInputTransform(TRUE, &rcSource, &rcDest);
}

void Magnifier::unmagnify() const {
    // Need some exception
    MagSetFullscreenTransform(1.0, 0, 0);
}

File: src/magnifier.rs
use windows::Win32::Foundation::*;
use windows::Win32::System::Threading::GetCurrentThreadId;
use windows::Win32::UI::Magnification::*;
use windows::Win32::UI::WindowsAndMessaging::*;

use crate::config::*;

const DEAD_ZONE_FACTOR: f32 = 0.1;

pub struct Magnifier {
    factor: f32,
}

impl Magnifier {
    pub fn new() -> Self {
        Self { factor: 1.0 }
    }

    pub fn do_action(&mut self, action: &Action) -> Result<(), String> {
        match action {
            Action::Set(factor) => self.set(*factor),
            Action::Add(factor) => self.add(*factor),
            Action::Toggle(factor) => self.toggle(*factor),
            Action::Exit => unsafe {
                PostThreadMessageA(GetCurrentThreadId(), WM_QUIT, WPARAM(0), LPARAM(0))
                    .as_bool()
                    .then_some(())
                    .ok_or("failed to send exit message")?
            },
        };
        self.update()?;

        Ok(())
    }

    fn set(&mut self, factor: f32) {
        self.factor = factor;
    }

    fn add(&mut self, factor: f32) {
        self.factor += factor;
        if self.factor < 1.0 {
            self.factor = 1.0;
        }
    }

    fn toggle(&mut self, factor: f32) {
        if self.factor == 1.0 {
            self.factor = factor;
        } else {
            self.factor = 1.0;
        }
    }

    pub fn update(&self) -> Result<(), String> {
        let mut point = POINT::default();
        let width;
        let height;

        unsafe {
            // fetch cursor pos info
            if GetCursorPos(&mut point).as_bool() == false {
                Err("failed to get cursor position")?
            }

            // fetch screen info
            width = GetSystemMetrics(SM_CXSCREEN) as f32;
            height = GetSystemMetrics(SM_CYSCREEN) as f32;
        }

        // linear calculations
        let mul_x = width * (1.0 - 1.0 / self.factor) / ((1.0 - 2.0 * DEAD_ZONE_FACTOR) * width);
        let sub_x = mul_x * DEAD_ZONE_FACTOR * width;
        let max_x = width * (1.0 - 1.0 / self.factor);

        let mul_y = height * (1.0 - 1.0 / self.factor) / ((1.0 - 2.0 * DEAD_ZONE_FACTOR) * height);
        let sub_y = mul_y * DEAD_ZONE_FACTOR * height;
        let max_y = height * (1.0 - 1.0 / self.factor);

        let x = (point.x as f32 * mul_x - sub_x).min(max_x).max(0.0) as i32;
        let y = (point.y as f32 * mul_y - sub_y).min(max_y).max(0.0) as i32;

        // for input transform
        let dst = RECT {
            left: 0,
            top: 0,
            right: width as i32,
            bottom: height as i32,
        };
        let src = RECT {
            left: x,
            top: y,
            right: x + (width / self.factor) as i32,
            bottom: y + (height / self.factor) as i32,
        };

        unsafe {
            if MagSetFullscreenTransform(self.factor, x, y).as_bool() == false {
                Err("failed to magnify")?
            }
            // MagSetInputTransform requires uiAcess privileges
            #[cfg(not(debug_assertions))]
            if MagSetInputTransform(TRUE, &src, &dst).as_bool() == false {
                Err("failed to handle input transform")?
            }
        }

        Ok(())
    }
}
